### DISCUSSION TITLE
Potential cursor pagination bug with `hasPrevPage` and `hasNextPage`

### DISCUSSION THREAD

**Discussion by stefansundin** [mikro-orm/mikro-orm#5154] [General]
Hello, I'm testing out the new cursor pagination in v6 (using 6.0.4) and I've encountered a potential bug. It might be intended but I think it should probably be changed even if it is intended.

I made a reproduction here: https://github.com/stefansundin/mikro-orm-reproduction/tree/cursor-hasPrevPage-hasNextPage-bug

The problem is when you are paginating "backwards" with `last` and `before`, when you hit the "last" page (i.e. you arrive back at the "first" page), `hasPrevPage` and `hasNextPage` will be reversed from what you expect.

Here's the relevant code from the reproduction:

```typescript
test('basic CRUD example', async () => {
  orm.em.create(User, { name: 'Foo', email: 'foo' });
  orm.em.create(User, { name: 'Bar', email: 'bar' });
  orm.em.create(User, { name: 'Baz', email: 'baz' });
  await orm.em.flush();
  orm.em.clear();

  // First stepping forward to the last item:

  const cursor1 = await orm.em.findByCursor(User, {}, { orderBy: { id: QueryOrder.ASC }, first: 1 });
  expect(cursor1.items[0].name).toBe('Foo');
  expect(cursor1.hasPrevPage).toBe(false);
  expect(cursor1.hasNextPage).toBe(true);

  const cursor2 = await orm.em.findByCursor(User, {}, { orderBy: { id: QueryOrder.ASC }, first: 1, after: cursor1 });
  expect(cursor2.items[0].name).toBe('Bar');
  expect(cursor2.hasPrevPage).toBe(true);
  expect(cursor2.hasNextPage).toBe(true);

  const cursor3 = await orm.em.findByCursor(User, {}, { orderBy: { id: QueryOrder.ASC }, first: 1, after: cursor2 });
  expect(cursor3.items[0].name).toBe('Baz');
  expect(cursor3.hasPrevPage).toBe(true);
  expect(cursor3.hasNextPage).toBe(false);

  // Then stepping backwards again:

  const cursor4 = await orm.em.findByCursor(User, {}, { orderBy: { id: QueryOrder.ASC }, last: 1, before: cursor3 });
  expect(cursor4.items[0].name).toBe('Bar');
  expect(cursor4.hasPrevPage).toBe(true);
  expect(cursor4.hasNextPage).toBe(true);

  const cursor5 = await orm.em.findByCursor(User, {}, { orderBy: { id: QueryOrder.ASC }, last: 1, before: cursor4 });
  expect(cursor5.items[0].name).toBe('Foo');
  // Here's the bug: hasPrevPage and hasNextPage are swapped!!
  expect(cursor5.hasPrevPage).toBe(false); // this is actually true
  expect(cursor5.hasNextPage).toBe(true); // this is actually false
});
```

I haven't used cursor pagination a ton in other ORMs or frameworks, so I don't know if this is how other projects work. But if it is intended then I would have to swap the booleans manually whenever I am paginating backwards.

As an attempt to see how others do it, I checked out how the GitHub graphql API works, and there it is behaving the way I expect. You can try this [in their graphql explorer](https://docs.github.com/en/graphql/overview/explorer):

```gql
query {
  organization(login: "github") {
    part1: repositories(first: 1) {
      __typename
      nodes {
        name
      }
      pageInfo {
        hasNextPage
        hasPreviousPage
        startCursor
        endCursor
      }
    }
    
    part2: repositories(first: 1, after:"Y3Vyc29yOnYyOpHNDJY=") {
      __typename
      nodes {
        name
      }
      pageInfo {
        hasNextPage
        hasPreviousPage
        startCursor
        endCursor
      }
    }
    
    part3: repositories(last: 1, before:"Y3Vyc29yOnYyOpHNPjk=") {
      __typename
      nodes {
        name
      }
      pageInfo {
        hasNextPage
        hasPreviousPage
        startCursor
        endCursor
      }
    }
  }
}
```

Here's the response I got:

```json
{
  "data": {
    "organization": {
      "part1": {
        "__typename": "RepositoryConnection",
        "nodes": [
          {
            "name": "media"
          }
        ],
        "pageInfo": {
          "hasNextPage": true,
          "hasPreviousPage": false,
          "startCursor": "Y3Vyc29yOnYyOpHNDJY=",
          "endCursor": "Y3Vyc29yOnYyOpHNDJY="
        }
      },
      "part2": {
        "__typename": "RepositoryConnection",
        "nodes": [
          {
            "name": "albino"
          }
        ],
        "pageInfo": {
          "hasNextPage": true,
          "hasPreviousPage": true,
          "startCursor": "Y3Vyc29yOnYyOpHNPjk=",
          "endCursor": "Y3Vyc29yOnYyOpHNPjk="
        }
      },
      "part3": {
        "__typename": "RepositoryConnection",
        "nodes": [
          {
            "name": "media"
          }
        ],
        "pageInfo": {
          "hasNextPage": true,
          "hasPreviousPage": false,
          "startCursor": "Y3Vyc29yOnYyOpHNDJY=",
          "endCursor": "Y3Vyc29yOnYyOpHNDJY="
        }
      }
    }
  }
}
```

In the last request, `hasPreviousPage` is `false` and `hasNextPage` is `true`, which makes sense.

I am opening this as a discussion as it might be intended. I know v6 is brand new so the docs are not fully fleshed out on the details, but I think eventually some of this behavior should be documented [here](https://mikro-orm.io/docs/entity-manager#cursor-based-pagination).

Thanks!!!

**Comment by stefansundin** [comment]
@B4nan Hi. Not sure if you saw this or not but I would love to hear your thoughts on it. Thank you!

**Reply by B4nan** [reply]
Sorry, I am out on vacation this week. Checked this briefly, and I'd say you are right that those two should be swapped. Sounds like an easy fix, PR welcome, I can have a look next week if it would be too hard for you.

**Reply by stefansundin** [reply]
No worries, I can wait. Sorry for the ping. Enjoy your vacation. 🏖️

**Comment by darkbasic** [comment]
The `hasNextPage`/`hasPreviousPage` logic should be swapped depending on whether you are requesting items before or after the cursor, or at least that's how I usually implement cursor based pagination.

That being said the third query feels weird to me: if `hasPreviousPage` is *false* and you ask for items **before** the first item it should return none. The fact that GitHub still returns the first element puzzles me, but probably past me has more knowledge about this and I should look at my own implementations to check how I used to implement this.

**Comment by darkbasic** [comment]
This is being implemented right now: https://github.com/mikro-orm/mikro-orm/pull/5320

**Reply by stefansundin** [reply]
Awesome. 👍 

I have been wanting to do more research to come up with more data that would back making this change, but I just haven't found the time.
